src/mpid/ch4/shm/src/shm_p2p.h

```diff
@@ -68,17 +68,37 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_SHM_mmods_try_matched_recv(void *buf,
 #ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
     /* XPMEM special receive */
     if (MPIDI_SHM_REQUEST(message, status) & MPIDI_SHM_REQ_XPMEM_SEND_LMT) {
+        MPIDI_XPMEM_am_unexp_rreq_t *unexp_rreq = &MPIDI_XPMEM_REQUEST(message, unexp_rreq);
+        MPIR_Comm *root_comm = NULL;
+        int recvtype_iscontig;
+
+        MPIR_Datatype_is_contig(datatype, &recvtype_iscontig);
+        if (recvtype_iscontig) {
+            MPIDI_XPMEM_REQUEST(message, counter) =
+                &MPIDI_XPMEM_global.coop_counter[MPIDI_XPMEM_global.local_rank *
+                                                 MPIDI_XPMEM_global.num_local +
+                                                 unexp_rreq->src_lrank];
+
+            root_comm = MPIDIG_context_id_to_comm(MPIDIG_REQUEST(message, context_id));
+
+            mpi_errno = MPIDI_XPMEM_lmt_cts_isend(buf, count, datatype,
+                                                  MPIDIG_REQUEST(message, rank), root_comm,
+                                                  &message);
+            MPIR_ERR_CHECK(mpi_errno);
```

Comment:
Does it makes more sense to put CTS response as a part of `handle_lmt_rts_recv`?

Besides, it is confusing to define `MPIDI_XPMEM_lmt_cts_isend` (called by receiver) and `MPIDI_XPMEM_lmt_rts_isend` (called by sender) in `xpmem_send.h`. `RTS` and `CTS` are just two internal handshake header types to accomplish the coop copy. I'd suggest:
- Rename `MPIDI_XPMEM_lmt_rts_isend` to `MPIDI_XPMEM_lmt_isend`, which means this is the send routine to initialize coop copy.
- Merge `MPIDI_XPMEM_lmt_cts_isend` into `MPIDI_XPMEM_handle_lmt_rts_recv`.

Reply:
I prefer to separate CTS and RTS routine when message is since current implementation can fall back to single-copy XPMEM in `MPIDI_XPMEM_handle_lmt_rts_recv` which will not call `MPIDI_XPMEM_lmt_cts_isend`. In the future, when we support all cases for cooperative copy we can put CTS routine into RTS.

Reply:
Let me try to understand your code design:
- Sender side: 
   - Initial: always issue `XPMEM_RTS` if send_dtype is contig (`MPIDI_XPMEM_lmt_isend` or `MPIDI_XPMEM_lmt_rts_isend` in your code); otherwise fallback to **POSIX send**.
   - Receive `XPMEM_CTS` (`MPIDI_XPMEM_ctrl_send_lmt_cts_cb`):
        - Perform **coop-copy**:
            1. start copy (`MPIDI_XPMEM_do_lmt_cooperative_copy`);
            2. reply `SEND_FIN`(`RECV_FIN` in your code) if it copies the last chunk.
    - Receive `FIN`: complete send request

- Receiver side: 
    - Receive `XPMEM_RTS` or initial receive (`MPIDI_XPMEM_ctrl_send_lmt_rts_cb` or `MPIDI_SHM_mmods_try_matched_recv`):
      - match a request
      - If matched, handle it<<define `MPIDI_XPMEM_handle_lmt_recv`?>>:
        - If recv_dtype is contig, perform **coop-copy** <<define `MPIDI_XPMEM_handle_lmt_coop_recv`?>>:
           1. reply `XPMEM_CTS` to sender;
           2. start copy (`MPIDI_XPMEM_do_lmt_cooperative_copy`) <<how about shorten to `MPIDI_XPMEM_do_lmt_coop_copy`?>>;
           3. reply `RECV_FIN`(`SEND_FIN` in your code) if it copies the last chunk.
        - Otherwise perform **single-copy** <<define `MPIDI_XPMEM_handle_lmt_single_recv`?>>:
           1. start copy;
           2. reply `RECV_FIN` to sender.
    - Receive `FIN`: complete receive request

Is the above breakdown correct? Please also see my suggestions marked with <<>>

Reply:
Sender side:
 - if sender copies the last chunk and send `RECV_FIN`, it will complete its request and not get `FIN` from receiver side. (similar to receiver side)

Reply:
Sure. I did not mean the sender should issue `RECV_FIN ` and also receive `FIN` from receiver.